checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -1113,34 +1173,71 @@ private void updateObligationsForAssignment(
+
+        boolean inConstructor =
+            TreePathUtil.inConstructor(typeFactory.getPath(assignmentNode.getTree()));
+
+        // Determine which obligations this field assignment can clear.  In a constructor,
+        // assignments to `this.field` only clears obligations on normal return, since
+        // on exception `this` becomes inaccessible.
+        Set<MethodExitKind> toClear;
+        if (inConstructor
+            && lhs instanceof FieldAccessNode
+            && ((FieldAccessNode) lhs).getReceiver() instanceof ThisNode) {
+          toClear = Collections.singleton(MethodExitKind.NORMAL_RETURN);
+        } else {
+          toClear = MethodExitKind.ALL;
+        }
+
+        @Nullable Element enclosingElem = lhsElement.getEnclosingElement();
+        @Nullable TypeElement enclosingType =
+            enclosingElem != null ? ElementUtils.enclosingTypeElement(enclosingElem) : null;
+
         // Assigning to an owning field is sufficient to clear a must-call alias obligation
         // in a constructor, if the enclosing class has at most one @Owning field. If the
         // class had multiple owning fields, then a soundness bug would occur: the must call
         // alias relationship would allow the whole class' obligation to be fulfilled by
         // closing only one of the parameters passed to the constructor (but the other
         // owning fields might not actually have had their obligations fulfilled). See test
         // case checker/tests/resourceleak/TwoOwningMCATest.java for an example.
-        Element enclosingCtr = lhsElement.getEnclosingElement();
-        if (enclosingCtr != null
-            && enclosingCtr.getKind() != ElementKind.CONSTRUCTOR
-            && hasAtMostOneOwningField(ElementUtils.enclosingTypeElement(enclosingCtr))) {
-          removeObligationsContainingVar(obligations, (LocalVariableNode) rhs);
+        if (hasAtMostOneOwningField(enclosingType)) {
+          removeObligationsContainingVar(
+              obligations, rhsVar, MustCallAliasHandling.NO_SPECIAL_HANDLING, toClear);
         } else {
-          removeObligationsContainingVarIfNotDerivedFromMustCallAlias(
-              obligations, (LocalVariableNode) rhs);
+          removeObligationsContainingVar(
+              obligations,
+              rhsVar,
+              MustCallAliasHandling.RETAIN_OBLIGATIONS_DERIVED_FROM_A_MUST_CALL_ALIAS_PARAMETER,
+              toClear);
+        }
+
+        // Finally, if any obligations containing this var remain, then closing the field will
+        // satisfy them.  Here we are overly cautious and only track final fields.  In the
+        // future we could perhaps relax this guard with careful handling for field reassignments.
+        if (ElementUtils.isFinal(lhsElement)) {
+          addAliasToObligationsContainingVar(
+              obligations,
+              rhsVar,
+              new ResourceAlias(JavaExpression.fromNode(lhs), lhsElement, lhs.getTree()));
```

Comment:
Is this shifting around some extant logic, or handling a new case?  If it's the latter, what is the motivation?

Reply:
This is handling a new case, and I'm glad you called it out because I should have mentioned it in the PR description.

Another big change here is that `ResourceAlias` can contain non-local-variable expressions. This is somewhat dangerous, so the only place that happens is right here, where final member fields are added to obligations' resource-alias sets.

The motivation is to support this test case in `OwnershipWithExceptions.java`:
```java
    FinalOwnedField(int ignored) throws IOException {
      // Same as the 0-argument constructor, but handled correctly (algorithm 1).
      resource = alloc();
      try {
        if (arbitraryChoice()) {
          throw new IOException();
        }
      } catch (Exception e) {
        resource.close();
        throw e;
      }
    }
```

Without this change, the call to `resource.close()` does not satisfy the RLC that the value returned by `alloc()` has been closed.